# <code style="background:yellow;color:black">Revision of last lecture:</code>

**Today we will be discussing Pfizer_1 dataset only. But there are a bunch of topics that we will cover today.** 

### <code style="background:yellow;color:black">Agenda for today:</code>

**We will cover little bit of datetime functions, string functions, but the majority of content for today is going to revolve around <code style="background:yellow;color:black">FILLING MISSING VALUES</code> as to how do we handle, manage and work if our dataframe contains missing values.**

In [1]:
import pandas as pd
import numpy as np

**Loading the dataset in pandas df.**

In [2]:
data = pd.read_csv("Pfizer_1.csv")

data

Date                Drug_Name    Parameter  1:30:00  2:30:00  \
0   15-10-2020  diltiazem hydrochloride  Temperature     23.0     22.0   
1   15-10-2020  diltiazem hydrochloride     Pressure     12.0     13.0   
2   15-10-2020      docetaxel injection  Temperature      NaN     17.0   
3   15-10-2020      docetaxel injection     Pressure      NaN     22.0   
4   15-10-2020   ketamine hydrochloride  Temperature     24.0      NaN   
5   15-10-2020   ketamine hydrochloride     Pressure      8.0      NaN   
6   16-10-2020  diltiazem hydrochloride  Temperature     34.0     35.0   
7   16-10-2020  diltiazem hydrochloride     Pressure     18.0     19.0   
8   16-10-2020      docetaxel injection  Temperature     46.0     47.0   
9   16-10-2020      docetaxel injection     Pressure     23.0     24.0   
10  16-10-2020   ketamine hydrochloride  Temperature      8.0      9.0   
11  16-10-2020   ketamine hydrochloride     Pressure     12.0     12.0   
12  17-10-2020  diltiazem hydrochloride  Temperature     20.0     19.0   
13  17-10-2020  diltiazem hydrochloride     Pressure      3.0      4.0   
14  17-10-2020      docetaxel injection  Temperature     12.0     13.0   
15  17-10-2020      docetaxel injection     Pressure     20.0     22.0   
16  17-10-2020   ketamine hydrochloride  Temperature     13.0     14.0   
17  17-10-2020   ketamine hydrochloride     Pressure      8.0      9.0   

    3:30:00  4:30:00  5:30:00  6:30:00  7:30:00  8:30:00  9:30:00  10:30:00  \
0       NaN     21.0     21.0       22     23.0     21.0     22.0        20   
1       NaN     11.0     13.0       14     16.0     16.0     24.0        18   
2      18.0      NaN     17.0       18      NaN      NaN     23.0        23   
3      22.0      NaN     22.0       23      NaN      NaN     27.0        26   
4       NaN     27.0      NaN       26     25.0     24.0     23.0        22   
5       NaN      7.0      NaN        9     10.0     11.0     10.0         9   
6      36.0     36.0     37.0       38     37.0     38.0     39.0        40   
7      20.0     21.0     22.0       23     24.0     25.0     25.0        24   
8       NaN     48.0     48.0       49     50.0     52.0     55.0        56   
9       NaN     25.0     26.0       27     28.0     29.0     28.0        28   
10     10.0      NaN     11.0       12     12.0     11.0      NaN        13   
11     13.0      NaN     15.0       15     15.0     15.0      NaN        16   
12     19.0     18.0     17.0       16     15.0      NaN     13.0        14   
13      4.0      4.0      6.0        8      9.0      NaN      9.0        11   
14     14.0     15.0     16.0       17     18.0     19.0     20.0        21   
15     22.0     22.0     22.0       23     25.0     26.0     27.0        28   
16     15.0     16.0     17.0       18     19.0     20.0     21.0        22   
17     10.0     11.0     11.0       12     12.0     11.0     12.0        13   

    11:30:00  12:30:00  
0       20.0        21  
1       19.0        20  
2       25.0        25  
3       29.0        28  
4       21.0        20  
5        9.0        11  
6        NaN        42  
7        NaN        27  
8       57.0        58  
9       29.0        30  
10      14.0        15  
11      17.0        18  
12      11.0        10  
13      13.0        14  
14      22.0        23  
15      29.0        28  
16      23.0        24  
17      14.0        15

**Checking the shape of our dataframe. It has 18 rows and 15 columns.**

In [3]:
data.shape

(18, 15)

* **Below we are doing <code style="background:yellow;color:black">melting of data.</code>**

* Below is our data_melt. Now as we can see its containing 216 rows and 5 columns in contrast to the shape of our data which was 18 rows and 15 columns. 

* We did this in the previous leture. 

In [4]:
data_melt = pd.melt(
    data,
    id_vars = ["Date", "Drug_Name", "Parameter"],
    var_name = "Time",
    value_name = "Reading"
)

data_melt

Date                Drug_Name    Parameter      Time  Reading
0    15-10-2020  diltiazem hydrochloride  Temperature   1:30:00     23.0
1    15-10-2020  diltiazem hydrochloride     Pressure   1:30:00     12.0
2    15-10-2020      docetaxel injection  Temperature   1:30:00      NaN
3    15-10-2020      docetaxel injection     Pressure   1:30:00      NaN
4    15-10-2020   ketamine hydrochloride  Temperature   1:30:00     24.0
..          ...                      ...          ...       ...      ...
211  17-10-2020  diltiazem hydrochloride     Pressure  12:30:00     14.0
212  17-10-2020      docetaxel injection  Temperature  12:30:00     23.0
213  17-10-2020      docetaxel injection     Pressure  12:30:00     28.0
214  17-10-2020   ketamine hydrochloride  Temperature  12:30:00     24.0
215  17-10-2020   ketamine hydrochloride     Pressure  12:30:00     15.0

[216 rows x 5 columns]

* In the previous lecture, we also tidy our data called <code style="background:yellow;color:black">data_tidy.</code> Below we are doing the same.

* Here we use <code style="background:yellow;color:black">pivot() method.</code>

* Our data_tidy has 108 rows and 2 columns because there are 3 levels of indexes in this dataframe.

In [5]:
data_tidy = data_melt.pivot(
    index = ["Date", "Time", "Drug_Name"],
    columns = "Parameter",
    values = "Reading"
)

data_tidy

Parameter                                    Pressure  Temperature
Date       Time     Drug_Name                                     
15-10-2020 10:30:00 diltiazem hydrochloride      18.0         20.0
                    docetaxel injection          26.0         23.0
                    ketamine hydrochloride        9.0         22.0
           11:30:00 diltiazem hydrochloride      19.0         20.0
                    docetaxel injection          29.0         25.0
...                                               ...          ...
17-10-2020 8:30:00  docetaxel injection          26.0         19.0
                    ketamine hydrochloride       11.0         20.0
           9:30:00  diltiazem hydrochloride       9.0         13.0
                    docetaxel injection          27.0         20.0
                    ketamine hydrochloride       12.0         21.0

[108 rows x 2 columns]

* This is what we do to fix our index in the data_tidy. 
* **Now we have a presentable single-indexed dataframe which contains 108 rows and 5 columns.**

In [6]:
data_tidy = data_tidy.reset_index()

data_tidy

Parameter        Date      Time                Drug_Name  Pressure  \
0          15-10-2020  10:30:00  diltiazem hydrochloride      18.0   
1          15-10-2020  10:30:00      docetaxel injection      26.0   
2          15-10-2020  10:30:00   ketamine hydrochloride       9.0   
3          15-10-2020  11:30:00  diltiazem hydrochloride      19.0   
4          15-10-2020  11:30:00      docetaxel injection      29.0   
..                ...       ...                      ...       ...   
103        17-10-2020   8:30:00      docetaxel injection      26.0   
104        17-10-2020   8:30:00   ketamine hydrochloride      11.0   
105        17-10-2020   9:30:00  diltiazem hydrochloride       9.0   
106        17-10-2020   9:30:00      docetaxel injection      27.0   
107        17-10-2020   9:30:00   ketamine hydrochloride      12.0   

Parameter  Temperature  
0                 20.0  
1                 23.0  
2                 22.0  
3                 20.0  
4                 25.0  
..                 ...  
103               19.0  
104               20.0  
105               13.0  
106               20.0  
107               21.0  

[108 rows x 5 columns]

**.columns is an attribute which shows all the columns in our dataframe and there is also a name which represents index name.**

In [7]:
data_tidy.columns

Index(['Date', 'Time', 'Drug_Name', 'Pressure', 'Temperature'], dtype='object', name='Parameter')

**Here we are setting the index name which is "parameter" to none.**

In [8]:
data_tidy.columns.name = None

data_tidy

Date      Time                Drug_Name  Pressure  Temperature
0    15-10-2020  10:30:00  diltiazem hydrochloride      18.0         20.0
1    15-10-2020  10:30:00      docetaxel injection      26.0         23.0
2    15-10-2020  10:30:00   ketamine hydrochloride       9.0         22.0
3    15-10-2020  11:30:00  diltiazem hydrochloride      19.0         20.0
4    15-10-2020  11:30:00      docetaxel injection      29.0         25.0
..          ...       ...                      ...       ...          ...
103  17-10-2020   8:30:00      docetaxel injection      26.0         19.0
104  17-10-2020   8:30:00   ketamine hydrochloride      11.0         20.0
105  17-10-2020   9:30:00  diltiazem hydrochloride       9.0         13.0
106  17-10-2020   9:30:00      docetaxel injection      27.0         20.0
107  17-10-2020   9:30:00   ketamine hydrochloride      12.0         21.0

[108 rows x 5 columns]

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">NaN and None values:</code>

**In pandas, <code style="background:yellow;color:black">NaN and None are used to represent missing or undefined values in a DataFrame or Series.</code> Ultimately both NaN and None both represent the same thing in terms of data and that is the value is missing or undefined. While they are both used to indicate the absence of data, there are some differences between them:**

**<code style="background:yellow;color:black">NaN (Not a Number):</code>**

* **NaN is a special floating-point value</code> defined by the IEEE floating-point standard. It is commonly used in pandas to represent missing or undefined numerical data.**
* When we perform mathematical operations involving NaN, the result is usually another NaN.
* In pandas, NaN is of the **float type.**

**<code style="background:yellow;color:black">None:</code>**

* **None is a Python object that represents the absence of a value or a null value.</code> It is a general-purpose object and can be used to represent missing values in both numerical and non-numerical (e.g., object, string) data.**
* When working with None, its behavior can vary depending on the operations.
* In pandas, None is often automatically replaced by NaN when working with numeric data.

**<code style="background:yellow;color:black">NaN -> Not a Number -> Value is missing or undefined</code>**<br>
**<code style="background:yellow;color:black">None -> Nothing -> Value is missing or undefined</code>**


* **So what pandas does that is if any column is numerical, then it will automatically keep the missing values as "NaN". And if the column is not numerical, string for example, then pandas will keep the missing values as "None".** 

* Pandas can use them interchangeably because it sees no difference between Nan and None. It is only for presentation purposes, internally they are aliases to one another. So we dont have to remember Nan or None, pandas has created functions which recognises both of them, all that we need to remember is that either NaN or None represent missing values. 

***
### <code style="background:yellow;color:black">Concept behind NaN and None values:</code>

* **<code style="background:yellow;color:black">"None"</code> means we can't call that there is nothing there.** 

* **There is this <code style="background:yellow;color:black">"placeholder value None"</code> which repressents that there is no value. But to represent no value python and pandas need some value which is kept there.**

***

* Lets take a very random example. Lets suppose, that we are in a movie theatre in which there is a special screening and there are no tickets, so its a first come first serve seating. 

* Lets say there is a group of 5 friends but only 3 of the people showed up and they are holding 2 more seats for 2 of their friends who will come later. 

* What they would do is either they would keep their bag or keep something of their own so that nobody comes and sits on the seat. 

* **<code style="background:yellow;color:black">Now that seat is in a gray area.</code> Physically it is empty since nobody is seating there, it can be taken, but some value (bag in our example) has been kept on top of the seat as a placeholder to occupy it so that it can signify that the seat is empty.** 

* **<code style="background:yellow;color:black">So it would be better if we focus on how to detect, deal, handle or how to act with them,</code> instead of the way they work in memory.**

***

* **The very reason of creating something like a pandas library is that <code style="background:yellow;color:black">we dont have to worry about what these NaN or None values are doing inside of memory.</code>** 

* **The very reason why we have all of this interface in front of us is that the memory layer remains hidden from us, otherwise we would have been coding in C language or Java. Because in C we get direct access to memory using some functions, we can see and work on that memory interface.**

* **Pandas dont give us access to that memory interface and the reason is <code style="background:yellow;color:black">having memory access and managing that memory is difficult.</code>** 

* Python has done certain memory optimizations in which 2 different things can share memory and python manages the shared memory. 

* Lets suppose if we store the number "5" in two different variables, python manages it in this way that it stores "5" only once in the memory and then there are two variables pointing to it. 

* **<code style="background:yellow;color:black">All of this optimization is done internally in python. If we were to do that, it would be very difficult for us. Similarly pandas has done a lot of optimizations on top of python.</code>** 

* **So having this capability in pandas is the very reason we are doing in pandas, otherwise we woud have directly done in C language.** 

In [9]:
data.head()

Date                Drug_Name    Parameter  1:30:00  2:30:00  \
0  15-10-2020  diltiazem hydrochloride  Temperature     23.0     22.0   
1  15-10-2020  diltiazem hydrochloride     Pressure     12.0     13.0   
2  15-10-2020      docetaxel injection  Temperature      NaN     17.0   
3  15-10-2020      docetaxel injection     Pressure      NaN     22.0   
4  15-10-2020   ketamine hydrochloride  Temperature     24.0      NaN   

   3:30:00  4:30:00  5:30:00  6:30:00  7:30:00  8:30:00  9:30:00  10:30:00  \
0      NaN     21.0     21.0       22     23.0     21.0     22.0        20   
1      NaN     11.0     13.0       14     16.0     16.0     24.0        18   
2     18.0      NaN     17.0       18      NaN      NaN     23.0        23   
3     22.0      NaN     22.0       23      NaN      NaN     27.0        26   
4      NaN     27.0      NaN       26     25.0     24.0     23.0        22   

   11:30:00  12:30:00  
0      20.0        21  
1      19.0        20  
2      25.0        25  
3      29.0        28  
4      21.0        20

In [10]:
type(None)

NoneType

In [11]:
type(np.nan)

float

* **To prove internal memory optimizations of python and pandas for NaN and None values, lets create a pandas series like this: pd.Series([1, np.nan, 2, None]).**

* Here there are three values of datatype "float" and one value of datatype "None".

* Now pandas has done internal optimization to convert NoneType to NaN (float), as we can see in the output.

In [12]:
pd.Series([1, np.nan, 2, None])

0    1.0
1    NaN
2    2.0
3    NaN
dtype: float64

* Lets suppose we have a pandas series like the one below, where 3 values are strings or string objects, 1 is Nan and 1 is None.

* For some reason, pandas keeps NaN as NaN and None as None, but it converts all to object type. 

* **So, this choice is internally made by pandas where it wants to convert NaN to None or vice versa and to which datatype. We are not in control of it. <code style="background:yellow;color:black">If its NaN or None, its not of the concern to us.</code>**

In [13]:
pd.Series(["1", "np.nan", "2", np.nan, None])

0         1
1    np.nan
2         2
3       NaN
4      None
dtype: object

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">isnull() and isna() functions:</code>

* **In Pandas, <code style="background:yellow;color:black">pd.isnull() and isna() functions</code> are used interchangeably to detect missing or null values in a DataFrame or Series. Both functions return a Boolean mask of the same shape as the input, where True indicates the presence of a missing or null value, and False indicates a non-null value.**

* isnull() and isna() are essentially aliases for the same function. They can be used interchangeably, and there is no difference in functionality between them. 

* The reason for pandas to keep aliases is to maintain consistency. Because earlier versions used to contain isnull() and now we have isna(), so since pandas want to maintain backward compatibility, they have both these aliases which are exactly the same thing.

* Below as we can see code cells "pd.isnull" and "pd.isna" return the same output, means they do exactly the same thing and are just aliases to one common function.

* Below to below code cell, "data.isna().head()" returns the dataframe in a True and False format. It tells us that the True values are the NaN or None values.

In [14]:
pd.isnull

<function pandas.core.dtypes.missing.isna(obj: 'object') -> 'bool | npt.NDArray[np.bool_] | NDFrame'>

In [15]:
pd.isna

<function pandas.core.dtypes.missing.isna(obj: 'object') -> 'bool | npt.NDArray[np.bool_] | NDFrame'>

In [16]:
data.isna().head()

Date  Drug_Name  Parameter  1:30:00  2:30:00  3:30:00  4:30:00  5:30:00  \
0  False      False      False    False    False     True    False    False   
1  False      False      False    False    False     True    False    False   
2  False      False      False     True    False    False     True    False   
3  False      False      False     True    False    False     True    False   
4  False      False      False    False     True     True    False     True   

   6:30:00  7:30:00  8:30:00  9:30:00  10:30:00  11:30:00  12:30:00  
0    False    False    False    False     False     False     False  
1    False    False    False    False     False     False     False  
2    False     True     True    False     False     False     False  
3    False     True     True    False     False     False     False  
4    False    False    False    False     False     False     False

* **By this code below, we can find out <code style="background:yellow;color:black">how many NaN or None values are there in each column.</code>** 

* We can do it using data.info() also. 

* How does this work? -- Value True corresponds to the number "1" and the value False corresponds to the number "0". So, this basically adds all the true values where the NaN values were. So it tells us that how many NaN or None values are there in each column. 

* There can be a possibility that some columns have so many None values and some columns might not have any None values.

* **Here we can find <code style="background:yellow;color:black">an insight that at 3:30 time, our drug is failing a lot.</code>** Maybe something is going wrong with the equipment. 

* If we want to find out how many NaN or None values are there in each and every row, we can do like the code implemented below. axis = 0 i.e. vertical axis is there by default.

In [17]:
data.isna().sum()

Date         0
Drug_Name    0
Parameter    0
1:30:00      2
2:30:00      2
3:30:00      6
4:30:00      4
5:30:00      2
6:30:00      0
7:30:00      2
8:30:00      4
9:30:00      2
10:30:00     0
11:30:00     2
12:30:00     0
dtype: int64

In [18]:
data.isna().sum(axis = 1)

0     1
1     1
2     4
3     4
4     3
5     3
6     1
7     1
8     1
9     1
10    2
11    2
12    1
13    1
14    0
15    0
16    0
17    0
dtype: int64

<hr style="border: 1px solid gray;">

# <code style="background:yellow;color:black">What to do with NaN/None values:</code>

* **Now what we need to do with missing/NaN/None values? There are two ways to handle them:**

* **First is to <code style="background:yellow;color:black">fill in these missing/NaN/None values</code> with some other values to make sure there is no missing value in our data. Or we might have to do "value approximation" in some way.** 

* **Second way could be that we <code style="background:yellow;color:black">drop all the rows and columns with missing values.</code>**

### <code style="background:yellow;color:black">dropna() method:</code>

* **The <code style="background:yellow;color:black">dropna() method</code> in Pandas is used to remove missing values (NaN or None) from a DataFrame or Series.** 

* **It is a method that allows us to eliminate rows or columns containing missing values.**

* By default axis = 0, i.e. horizontal axis. If we do axis = 1 in dropna() method, all the columns are deleted where the values are missing.

* **But by doing these deletions in code cell below, so much of data has been lost. We had 18 rows, now we have just 4 rows remaining, and we had 15 columns, now we have just 6 columns remaining. So this method made us loose too much data.**

* **So now we need to figure out a way to deal with missing values but not by deleting those rows and columns with missing values, but by filling in some values in their place.**

In [19]:
data.dropna()

Date               Drug_Name    Parameter  1:30:00  2:30:00  \
14  17-10-2020     docetaxel injection  Temperature     12.0     13.0   
15  17-10-2020     docetaxel injection     Pressure     20.0     22.0   
16  17-10-2020  ketamine hydrochloride  Temperature     13.0     14.0   
17  17-10-2020  ketamine hydrochloride     Pressure      8.0      9.0   

    3:30:00  4:30:00  5:30:00  6:30:00  7:30:00  8:30:00  9:30:00  10:30:00  \
14     14.0     15.0     16.0       17     18.0     19.0     20.0        21   
15     22.0     22.0     22.0       23     25.0     26.0     27.0        28   
16     15.0     16.0     17.0       18     19.0     20.0     21.0        22   
17     10.0     11.0     11.0       12     12.0     11.0     12.0        13   

    11:30:00  12:30:00  
14      22.0        23  
15      29.0        28  
16      23.0        24  
17      14.0        15

In [20]:
data.dropna(axis = 1)

Date                Drug_Name    Parameter  6:30:00  10:30:00  \
0   15-10-2020  diltiazem hydrochloride  Temperature       22        20   
1   15-10-2020  diltiazem hydrochloride     Pressure       14        18   
2   15-10-2020      docetaxel injection  Temperature       18        23   
3   15-10-2020      docetaxel injection     Pressure       23        26   
4   15-10-2020   ketamine hydrochloride  Temperature       26        22   
5   15-10-2020   ketamine hydrochloride     Pressure        9         9   
6   16-10-2020  diltiazem hydrochloride  Temperature       38        40   
7   16-10-2020  diltiazem hydrochloride     Pressure       23        24   
8   16-10-2020      docetaxel injection  Temperature       49        56   
9   16-10-2020      docetaxel injection     Pressure       27        28   
10  16-10-2020   ketamine hydrochloride  Temperature       12        13   
11  16-10-2020   ketamine hydrochloride     Pressure       15        16   
12  17-10-2020  diltiazem hydrochloride  Temperature       16        14   
13  17-10-2020  diltiazem hydrochloride     Pressure        8        11   
14  17-10-2020      docetaxel injection  Temperature       17        21   
15  17-10-2020      docetaxel injection     Pressure       23        28   
16  17-10-2020   ketamine hydrochloride  Temperature       18        22   
17  17-10-2020   ketamine hydrochloride     Pressure       12        13   

    12:30:00  
0         21  
1         20  
2         25  
3         28  
4         20  
5         11  
6         42  
7         27  
8         58  
9         30  
10        15  
11        18  
12        10  
13        14  
14        23  
15        28  
16        24  
17        15

### <code style="background:yellow;color:black">fillna() method:</code>

* **The <code style="background:yellow;color:black">fillna() method</code> in Pandas is used to fill missing or NaN (Not a Number) values with specified values. It can be applied to both DataFrames and Series to replace missing values with a constant, a calculated value, or values from nearby elements.**

* One way of filling the missing values is that we fill all those places with "0" like the code below. All missing values will now become 0. 

* We pass the argument 0 in the fillna() method. We can check below that every single missing value has been filled with 0.

* There is a problem with filling in the missing values with 0. Lets suppose we are measuring pressure for 5 consecutive hours.  We will encounter incorrect average with 0s in the place of missing values. And when there are a lot of 0s, the average can further be more incorrect. REFER NOTES HERE FOR MORE CLARITY.

* **So filling missing values with 0s is not right or not the best way.**

* **Then shall we do some sort of estimator like <code style="background:yellow;color:black">mean, median or mode</code>? -- Yes thats possible.** 

In [21]:
data.fillna(0).head()

Date                Drug_Name    Parameter  1:30:00  2:30:00  \
0  15-10-2020  diltiazem hydrochloride  Temperature     23.0     22.0   
1  15-10-2020  diltiazem hydrochloride     Pressure     12.0     13.0   
2  15-10-2020      docetaxel injection  Temperature      0.0     17.0   
3  15-10-2020      docetaxel injection     Pressure      0.0     22.0   
4  15-10-2020   ketamine hydrochloride  Temperature     24.0      0.0   

   3:30:00  4:30:00  5:30:00  6:30:00  7:30:00  8:30:00  9:30:00  10:30:00  \
0      0.0     21.0     21.0       22     23.0     21.0     22.0        20   
1      0.0     11.0     13.0       14     16.0     16.0     24.0        18   
2     18.0      0.0     17.0       18      0.0      0.0     23.0        23   
3     22.0      0.0     22.0       23      0.0      0.0     27.0        26   
4      0.0     27.0      0.0       26     25.0     24.0     23.0        22   

   11:30:00  12:30:00  
0      20.0        21  
1      19.0        20  
2      25.0        25  
3      29.0        28  
4      21.0        20

**Checking that column "2:30:00"**

In [22]:
data["2:30:00"]

0     22.0
1     13.0
2     17.0
3     22.0
4      NaN
5      NaN
6     35.0
7     19.0
8     47.0
9     24.0
10     9.0
11    12.0
12    19.0
13     4.0
14    13.0
15    22.0
16    14.0
17     9.0
Name: 2:30:00, dtype: float64

* Lets suppose we were filling data for only one column i.e. "2:30:00". 
* **We can calculate the <code style="background:yellow;color:black">mean of this column using .mean() method.</code>** 
* We will fill in any missing value in this column with the mean of this column. And the mean of this column is 18.8125

In [23]:
data["2:30:00"].mean()

18.8125

* **Here we are filling the missing values of the column with the mean of the colum by using the code below-**

* In fillna() we are passing selected column .mean(). **

* **Now we have filled all the missing values of that column with the average or mean value of that column.**

* Here we have done a mistake. In one single column we have temperature and pressure values both and then there are temperature and pressure values for multiple drugs also. What we have done is we have taken temperature and pressure values for multiple drugs and taken an average of it. 

* **So our ideas was right its just that we have to calculate the average for temperature for one drug across all the times i.e. columns and fill that mean value there. So average have to be applied here in our case row-wise.**

* **But to write "mean" code for row in this horizontal data structure is cumbersome, and that is why for this analysis, we will be moving towards melted data. Thats why we need a correctly structured data, because it would make our analysis much much easier.**

In [24]:
data["2:30:00"].fillna(data["2:30:00"].mean())

0     22.0000
1     13.0000
2     17.0000
3     22.0000
4     18.8125
5     18.8125
6     35.0000
7     19.0000
8     47.0000
9     24.0000
10     9.0000
11    12.0000
12    19.0000
13     4.0000
14    13.0000
15    22.0000
16    14.0000
17     9.0000
Name: 2:30:00, dtype: float64

** To solve this issue we will be using data_tidy, apply() and groupby() in some form. **

In [25]:
data_tidy

Date      Time                Drug_Name  Pressure  Temperature
0    15-10-2020  10:30:00  diltiazem hydrochloride      18.0         20.0
1    15-10-2020  10:30:00      docetaxel injection      26.0         23.0
2    15-10-2020  10:30:00   ketamine hydrochloride       9.0         22.0
3    15-10-2020  11:30:00  diltiazem hydrochloride      19.0         20.0
4    15-10-2020  11:30:00      docetaxel injection      29.0         25.0
..          ...       ...                      ...       ...          ...
103  17-10-2020   8:30:00      docetaxel injection      26.0         19.0
104  17-10-2020   8:30:00   ketamine hydrochloride      11.0         20.0
105  17-10-2020   9:30:00  diltiazem hydrochloride       9.0         13.0
106  17-10-2020   9:30:00      docetaxel injection      27.0         20.0
107  17-10-2020   9:30:00   ketamine hydrochloride      12.0         21.0

[108 rows x 5 columns]

* First we groupby the Drug_Name in the data_tidy, then on top of that we calculate the average for temperature.
* Secondly we fill in that average into the missing values. 

* Step 1- Groupby the drug name, calculate the average of temperature.
* Step2- Put these averages into the seperate column called temperature average, and then we will fill missing values using the temperature average column

* So that  temperature average column is going to contain the respective average for that paticular drug across all the times. For this we are going to use apply() method.

* In the function temp_mean(x), "x" is a group and in that entire group, we are adding a new column called "Temp_Avg" of which the column value is equal to the temperature column and the average of it. This is our apply function.

In [26]:
def temp_mean(x):
    # x will be a group
    # In that group, we are setting the value of the entire Temp_Avg column to the average value of the Temperature column
    
    x["Temp_Avg"] = x["Temperature"].mean()
    
    return x

We will groupby data_tidy on Drug_Name and then apply apply() method and in that we will call the temp_mean function defined above. "group_keys = False" is just to avoid the jupyter warning. 

In [27]:
data_tidy = data_tidy.groupby(["Drug_Name"], group_keys = False).apply(temp_mean)

data_tidy

Date      Time                Drug_Name  Pressure  Temperature  \
0    15-10-2020  10:30:00  diltiazem hydrochloride      18.0         20.0   
1    15-10-2020  10:30:00      docetaxel injection      26.0         23.0   
2    15-10-2020  10:30:00   ketamine hydrochloride       9.0         22.0   
3    15-10-2020  11:30:00  diltiazem hydrochloride      19.0         20.0   
4    15-10-2020  11:30:00      docetaxel injection      29.0         25.0   
..          ...       ...                      ...       ...          ...   
103  17-10-2020   8:30:00      docetaxel injection      26.0         19.0   
104  17-10-2020   8:30:00   ketamine hydrochloride      11.0         20.0   
105  17-10-2020   9:30:00  diltiazem hydrochloride       9.0         13.0   
106  17-10-2020   9:30:00      docetaxel injection      27.0         20.0   
107  17-10-2020   9:30:00   ketamine hydrochloride      12.0         21.0   

      Temp_Avg  
0    24.848485  
1    30.387097  
2    17.709677  
3    24.848485  
4    30.387097  
..         ...  
103  30.387097  
104  17.709677  
105  24.848485  
106  30.387097  
107  17.709677  

[108 rows x 6 columns]

In [28]:
data_tidy = data_tidy.sort_values(by = "Temp_Avg", ascending = False)

data_tidy

Date      Time               Drug_Name  Pressure  Temperature  \
55   16-10-2020   4:30:00     docetaxel injection      25.0         48.0   
79   17-10-2020  12:30:00     docetaxel injection      28.0         23.0   
46   16-10-2020   1:30:00     docetaxel injection      23.0         46.0   
22   15-10-2020   5:30:00     docetaxel injection      22.0         17.0   
85   17-10-2020   2:30:00     docetaxel injection      22.0         13.0   
..          ...       ...                     ...       ...          ...   
29   15-10-2020   7:30:00  ketamine hydrochloride      10.0         25.0   
80   17-10-2020  12:30:00  ketamine hydrochloride      15.0         24.0   
56   16-10-2020   4:30:00  ketamine hydrochloride       NaN          NaN   
26   15-10-2020   6:30:00  ketamine hydrochloride       9.0         26.0   
107  17-10-2020   9:30:00  ketamine hydrochloride      12.0         21.0   

      Temp_Avg  
55   30.387097  
79   30.387097  
46   30.387097  
22   30.387097  
85   30.387097  
..         ...  
29   17.709677  
80   17.709677  
56   17.709677  
26   17.709677  
107  17.709677  

[108 rows x 6 columns]

**So if we now look at data_tidy, what we have various columns including Temp_Avg. So if we look at this particular <mark>"docetaxel injection"</mark> group, the temperature average for it is <mark>30.387097</mark> and so on.** 	

In [30]:
data_tidy = data_tidy.reset_index()

data_tidy

index        Date      Time               Drug_Name  Pressure  \
0       55  16-10-2020   4:30:00     docetaxel injection      25.0   
1       79  17-10-2020  12:30:00     docetaxel injection      28.0   
2       46  16-10-2020   1:30:00     docetaxel injection      23.0   
3       22  15-10-2020   5:30:00     docetaxel injection      22.0   
4       85  17-10-2020   2:30:00     docetaxel injection      22.0   
..     ...         ...       ...                     ...       ...   
103     29  15-10-2020   7:30:00  ketamine hydrochloride      10.0   
104     80  17-10-2020  12:30:00  ketamine hydrochloride      15.0   
105     56  16-10-2020   4:30:00  ketamine hydrochloride       NaN   
106     26  15-10-2020   6:30:00  ketamine hydrochloride       9.0   
107    107  17-10-2020   9:30:00  ketamine hydrochloride      12.0   

     Temperature   Temp_Avg  
0           48.0  30.387097  
1           23.0  30.387097  
2           46.0  30.387097  
3           17.0  30.387097  
4           13.0  30.387097  
..           ...        ...  
103         25.0  17.709677  
104         24.0  17.709677  
105          NaN  17.709677  
106         26.0  17.709677  
107         21.0  17.709677  

[108 rows x 7 columns]

**We have calculated the average temperature but we have not filled the missing values. We will fill them by picking the corresponding average value. Below is the code for that-** 

In [31]:
data_tidy["Temperature"].fillna(data_tidy["Temp_Avg"], inplace = True)

**Output of this code below is 0, which means all our NaN or None values of Temperature column have been filled with corresponding Temp_Avg column values.**

In [33]:
data_tidy["Temperature"].isna().sum()

0

**There are 3 steps that we followed and will again follow for pressure-**

* <mark>Step 1</mark>- Write down an apply function.
* <mark>Step 2</mark>- Groupby the data and apply the apply() function to the fresh column 
* <mark>Step 3</mark>- Fill in the missing NaN or None values with corresponding average column values.

In [34]:
# STEP 1

def pressure_mean(x):
    # x will be a group
    # In that group, we are setting the value of the entire Pressure_Avg column to the average value of the Pressure column
    
    x["Pressure_Avg"] = x["Pressure"].mean()
    
    return x

In [35]:
# STEP 2

data_tidy = data_tidy.groupby(["Drug_Name"], group_keys = False).apply(pressure_mean)

In [36]:
# STEP 3

data_tidy["Pressure"].fillna(data_tidy["Pressure_Avg"], inplace = True)

**This code below outputs 0, which means all our NaN or None values of Pressure column have been filled with corresponding Pressure_Avg column values.**

In [37]:
data_tidy["Pressure"].isna().sum()

0

**This code below outputs 0, which means all our NaN or None values of Temperature and Pressure column have been filled with their  corresponding Temp_Avg and Pressure_Avg column values.**

**<mark>Our entire process of filling in the missing values is now COMPLETE.</mark>**

In [38]:
data_tidy.isna().sum()

index           0
Date            0
Time            0
Drug_Name       0
Pressure        0
Temperature     0
Temp_Avg        0
Pressure_Avg    0
dtype: int64

** Checking the dataframe. **

In [39]:
data_tidy

index        Date      Time               Drug_Name   Pressure  \
0       55  16-10-2020   4:30:00     docetaxel injection  25.000000   
1       79  17-10-2020  12:30:00     docetaxel injection  28.000000   
2       46  16-10-2020   1:30:00     docetaxel injection  23.000000   
3       22  15-10-2020   5:30:00     docetaxel injection  22.000000   
4       85  17-10-2020   2:30:00     docetaxel injection  22.000000   
..     ...         ...       ...                     ...        ...   
103     29  15-10-2020   7:30:00  ketamine hydrochloride  10.000000   
104     80  17-10-2020  12:30:00  ketamine hydrochloride  15.000000   
105     56  16-10-2020   4:30:00  ketamine hydrochloride  11.935484   
106     26  15-10-2020   6:30:00  ketamine hydrochloride   9.000000   
107    107  17-10-2020   9:30:00  ketamine hydrochloride  12.000000   

     Temperature   Temp_Avg  Pressure_Avg  
0      48.000000  30.387097     25.483871  
1      23.000000  30.387097     25.483871  
2      46.000000  30.387097     25.483871  
3      17.000000  30.387097     25.483871  
4      13.000000  30.387097     25.483871  
..           ...        ...           ...  
103    25.000000  17.709677     11.935484  
104    24.000000  17.709677     11.935484  
105    17.709677  17.709677     11.935484  
106    26.000000  17.709677     11.935484  
107    21.000000  17.709677     11.935484  

[108 rows x 8 columns]

* There are a bunch of temperatures here. We want to change this **continuous data** to **categorical data**. Here pressure and temperature is continuous data which means it can take any number of values. Whereas drug name is categorical value because it can take only certain values or names. 

* But lets suppose we want to convert temperature into a fresh column which contains some categorical form of temperatures, for example, we can have 4 categories like **LOW, MEDIUM, HIGH and VERY HIGH.**

* To do this, first we need to know the range of temperature values from lowest to highest. Below code lets us know that range. **So we now know that the minimum temperature is 8.0 and the maximum temperature is 58.0.**

* Lets assume that our temperature range is 5 - 60, and we accordingly want to form buckets or categories as discussed above. 

* We are creating <mark>4 random buckets.</mark> 

In [41]:
print(data_tidy["Temperature"].min(), data_tidy["Temperature"].max())

8.0 58.0


* We call these categories or buckets as "temp_points" as shown below.

* How will it form buckets? -- 5 will be low point and 20 will be the high point. Then 20 will be low point and 35 will be the high point and so on and so forth. We will make them a python list.

* Another thing that we create is the labels called as "temp_labels" for these buckets. We will make them a python list.

In [42]:
temp_points = [5, 20, 35, 50, 60]

temp_labels = ["low", "medium", "high", "very high"]

* **We use a new function called pd.cut(). It basically cuts or divides the continuous data into multiple categories.** 

* Lets suppose we create a new column called "Temp_Category". 

* **In pd.cut(), we will first pass the column on which we want to perform the categories. bins will be "temp_points", and labels will be "temp_labels".** And we are done.

In [44]:
data_tidy["Temp_Category"] = pd.cut(
    data_tidy["Temperature"],
    bins = temp_points,
    labels = temp_labels
)

data_tidy

index        Date      Time               Drug_Name   Pressure  \
0       55  16-10-2020   4:30:00     docetaxel injection  25.000000   
1       79  17-10-2020  12:30:00     docetaxel injection  28.000000   
2       46  16-10-2020   1:30:00     docetaxel injection  23.000000   
3       22  15-10-2020   5:30:00     docetaxel injection  22.000000   
4       85  17-10-2020   2:30:00     docetaxel injection  22.000000   
..     ...         ...       ...                     ...        ...   
103     29  15-10-2020   7:30:00  ketamine hydrochloride  10.000000   
104     80  17-10-2020  12:30:00  ketamine hydrochloride  15.000000   
105     56  16-10-2020   4:30:00  ketamine hydrochloride  11.935484   
106     26  15-10-2020   6:30:00  ketamine hydrochloride   9.000000   
107    107  17-10-2020   9:30:00  ketamine hydrochloride  12.000000   

     Temperature   Temp_Avg  Pressure_Avg Temp_Category  
0      48.000000  30.387097     25.483871          high  
1      23.000000  30.387097     25.483871        medium  
2      46.000000  30.387097     25.483871          high  
3      17.000000  30.387097     25.483871           low  
4      13.000000  30.387097     25.483871           low  
..           ...        ...           ...           ...  
103    25.000000  17.709677     11.935484        medium  
104    24.000000  17.709677     11.935484        medium  
105    17.709677  17.709677     11.935484           low  
106    26.000000  17.709677     11.935484        medium  
107    21.000000  17.709677     11.935484        medium  

[108 rows x 9 columns]

**Seeing the number of values in each category can give us some valuable insights. We might want to look at the reasons for the 5 values of drugs in the very high category which are reaching this high values often.** 

In [45]:
data_tidy["Temp_Category"].value_counts()

low          50
medium       38
high         15
very high     5
Name: Temp_Category, dtype: int64

**There are a bunch of string functions that are available to us while working with pandas.** 
 
* In python, we have the "in" operator to fetch a substring from a string. 
* In pandas too, there is a list of string functions, of which only one or two will be taught, rest we can google it.
* Whatever functions are available in pandas are there in this format -- <mark>.str.contains</mark> 
* **This will be the usual format that these functions will follow -- <mark>Series.str.function_name</mark>**<br><br>

* Lets say we want to fetch data for all the drug names which contain the name "hydrochloride" in the Drug_Name column. We can do that using the code below. It only prints the rows where the hydrochloride is present.

In [54]:
data_final.loc[data_final["Drug_Name"].str.contains("hydrochloride")]

index        Date      Time                Drug_Name   Pressure  \
36      57  16-10-2020   5:30:00  diltiazem hydrochloride  22.000000   
37      63  16-10-2020   7:30:00  diltiazem hydrochloride  24.000000   
38      66  16-10-2020   8:30:00  diltiazem hydrochloride  25.000000   
39      60  16-10-2020   6:30:00  diltiazem hydrochloride  23.000000   
40       0  15-10-2020  10:30:00  diltiazem hydrochloride  18.000000   
..     ...         ...       ...                      ...        ...   
103     29  15-10-2020   7:30:00   ketamine hydrochloride  10.000000   
104     80  17-10-2020  12:30:00   ketamine hydrochloride  15.000000   
105     56  16-10-2020   4:30:00   ketamine hydrochloride  11.935484   
106     26  15-10-2020   6:30:00   ketamine hydrochloride   9.000000   
107    107  17-10-2020   9:30:00   ketamine hydrochloride  12.000000   

     Temperature   Temp_Avg  Pressure_Avg Temp_Category  
36     37.000000  24.848485     15.424242          high  
37     37.000000  24.848485     15.424242          high  
38     38.000000  24.848485     15.424242          high  
39     38.000000  24.848485     15.424242          high  
40     20.000000  24.848485     15.424242           low  
..           ...        ...           ...           ...  
103    25.000000  17.709677     11.935484        medium  
104    24.000000  17.709677     11.935484        medium  
105    17.709677  17.709677     11.935484           low  
106    26.000000  17.709677     11.935484        medium  
107    21.000000  17.709677     11.935484        medium  

[72 rows x 9 columns]

** Lets take another example. The Date column we have is a string datatype and in the following format -- <mark>dd-mm-yyyy<mark> ** 

In [55]:
data_final["Date"]

0      16-10-2020
1      17-10-2020
2      16-10-2020
3      15-10-2020
4      17-10-2020
          ...    
103    15-10-2020
104    17-10-2020
105    16-10-2020
106    15-10-2020
107    17-10-2020
Name: Date, Length: 108, dtype: object

**By using the .str.split("-") method we are changing the format of the date and creating a list of individual date elements, which are of string datatype.**

In [56]:
data_final["Date"].str.split("-")

0      [16, 10, 2020]
1      [17, 10, 2020]
2      [16, 10, 2020]
3      [15, 10, 2020]
4      [17, 10, 2020]
            ...      
103    [15, 10, 2020]
104    [17, 10, 2020]
105    [16, 10, 2020]
106    [15, 10, 2020]
107    [17, 10, 2020]
Name: Date, Length: 108, dtype: object

* Lets say we want to only take out the <mark>year component</mark> from the Date. 
* In the below code, we are selecting the date column, splitting its string into a list of individual components and apply apply() method with a lambda function on top of it to fetch the year part.

In [57]:
data_final["Date"].str.split("-").apply(lambda x: x[2])

0      2020
1      2020
2      2020
3      2020
4      2020
       ... 
103    2020
104    2020
105    2020
106    2020
107    2020
Name: Date, Length: 108, dtype: object

* **Here we notice that in all above cases, datatype is an object i.e. string. What if we dont want the string, we want a proper date datatype. Is that possible? -- YES**

* In pandas, there are various functions which help us achieve that. 

* First of all lets create a new column called "Timestamp" which is a concatenation of Date and Time columns. We can do that using the below code.

* Second we will delete the Date and Time columns.

In [58]:
data_final["Timestamp"] = data_final["Date"] + " " + data_final["Time"]

In [59]:
data_final.drop(["Date", "Time"], axis = 1, inplace = True)

In [60]:
data_final

index               Drug_Name   Pressure  Temperature   Temp_Avg  \
0       55     docetaxel injection  25.000000    48.000000  30.387097   
1       79     docetaxel injection  28.000000    23.000000  30.387097   
2       46     docetaxel injection  23.000000    46.000000  30.387097   
3       22     docetaxel injection  22.000000    17.000000  30.387097   
4       85     docetaxel injection  22.000000    13.000000  30.387097   
..     ...                     ...        ...          ...        ...   
103     29  ketamine hydrochloride  10.000000    25.000000  17.709677   
104     80  ketamine hydrochloride  15.000000    24.000000  17.709677   
105     56  ketamine hydrochloride  11.935484    17.709677  17.709677   
106     26  ketamine hydrochloride   9.000000    26.000000  17.709677   
107    107  ketamine hydrochloride  12.000000    21.000000  17.709677   

     Pressure_Avg Temp_Category            Timestamp  
0       25.483871          high   16-10-2020 4:30:00  
1       25.483871        medium  17-10-2020 12:30:00  
2       25.483871          high   16-10-2020 1:30:00  
3       25.483871           low   15-10-2020 5:30:00  
4       25.483871           low   17-10-2020 2:30:00  
..            ...           ...                  ...  
103     11.935484        medium   15-10-2020 7:30:00  
104     11.935484        medium  17-10-2020 12:30:00  
105     11.935484           low   16-10-2020 4:30:00  
106     11.935484        medium   15-10-2020 6:30:00  
107     11.935484        medium   17-10-2020 9:30:00  

[108 rows x 8 columns]

* What we will do is we will convert this Timestamp datatype to a proper "datetime" datatype. 

* **There is a special function in pandas called <mark>to_datetime()</mark> to convert any datatype to "datetime" datatype.**

* **When we look into data_final.info(), we can see the datatype of Timestamp is "datatime".** The datatype has been changed from "object" to "datatime".

In [61]:
data_final["Timestamp"] = pd.to_datetime(data_final["Timestamp"])

In [62]:
data_final

index               Drug_Name   Pressure  Temperature   Temp_Avg  \
0       55     docetaxel injection  25.000000    48.000000  30.387097   
1       79     docetaxel injection  28.000000    23.000000  30.387097   
2       46     docetaxel injection  23.000000    46.000000  30.387097   
3       22     docetaxel injection  22.000000    17.000000  30.387097   
4       85     docetaxel injection  22.000000    13.000000  30.387097   
..     ...                     ...        ...          ...        ...   
103     29  ketamine hydrochloride  10.000000    25.000000  17.709677   
104     80  ketamine hydrochloride  15.000000    24.000000  17.709677   
105     56  ketamine hydrochloride  11.935484    17.709677  17.709677   
106     26  ketamine hydrochloride   9.000000    26.000000  17.709677   
107    107  ketamine hydrochloride  12.000000    21.000000  17.709677   

     Pressure_Avg Temp_Category           Timestamp  
0       25.483871          high 2020-10-16 04:30:00  
1       25.483871        medium 2020-10-17 12:30:00  
2       25.483871          high 2020-10-16 01:30:00  
3       25.483871           low 2020-10-15 05:30:00  
4       25.483871           low 2020-10-17 02:30:00  
..            ...           ...                 ...  
103     11.935484        medium 2020-10-15 07:30:00  
104     11.935484        medium 2020-10-17 12:30:00  
105     11.935484           low 2020-10-16 04:30:00  
106     11.935484        medium 2020-10-15 06:30:00  
107     11.935484        medium 2020-10-17 09:30:00  

[108 rows x 8 columns]

In [63]:
data_final.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 108 entries, 0 to 107
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype         
---  ------         --------------  -----         
 0   index          108 non-null    int64         
 1   Drug_Name      108 non-null    object        
 2   Pressure       108 non-null    float64       
 3   Temperature    108 non-null    float64       
 4   Temp_Avg       108 non-null    float64       
 5   Pressure_Avg   108 non-null    float64       
 6   Temp_Category  108 non-null    category      
 7   Timestamp      108 non-null    datetime64[ns]
dtypes: category(1), datetime64[ns](1), float64(4), int64(1), object(1)
memory usage: 11.1+ KB


* **What is the purpose of coverting into "datetime" datatype? -- We can extract individual components of the timestamp.**

* We are creating a seperate variable called "t" to extract differnt timestamp data as done below.

In [64]:
t = data_final["Timestamp"][0]

In [65]:
t

Timestamp('2020-10-16 04:30:00')

In [66]:
t.year

2020

In [67]:
t.month

10

In [68]:
t.day

16

In [69]:
t.month_name()

'October'

In [70]:
t.week

42

In [71]:
t.date()

datetime.date(2020, 10, 16)

In [72]:
t

Timestamp('2020-10-16 04:30:00')

* **There is something called as "strftime" or string format time. It is basically string representattions of various formats of datetime datatype. 

* There are a lot of representations and we do not need to memorize those. We can read the documentation or google these.

* Link to documentation- https://pandas.pydata.org/docs/reference/api/pandas.Period.strftime.html

In [73]:
t.strftime("%D")

'10/16/20'

** We can have any type of formatting, and these %d, %m, %y are sort of like placeholders. There are different placeholders for different components of the Timestamp, a seperate placeholder for date, a seperate placeholder for month and so on. **

In [74]:
t.strftime("%d-    %m-     %y")

'16-    10-     20'

* **<mark>CRUX</mark> of it is that we can convert a particular Date or Time column to a proper "datetime" datatype column.** 
* **We can fetch out different parts of Timestamp and we can format these parts of Timestamp as string we can do this using .strftime.**

**Any dataframe that we create, we start with loading the dataframe in pandas from a csv file. We can send back the dataframe to the csv file by the <mark>to_csv()</mark> method, as implemented below.** The new file "Pfizer_Final.csv" has been now saved in the same folder as this notebook.

In [75]:
data_final.to_csv("Pfizer_Final.csv", sep = ",", index = False)

* **Usually what will happen in real life is that we will be using SQL to get data and python to pre-process that data.** 

* Lets suppose you are working as an analyst, and if the data source gives you incorrect data, then also **you cant fix the data.** <mark>All we can do is fill in missing values, do data pre-processing, do approximations, do aggregations BUT we cant go back and change the data.</mark> **The original data will always remain the original data.** 

* Lets suppose you logged into scaler, and in that form filling you write your name as "UltimateLord554". Now Scaler cant change your name just because you filled incorrect name. Scaler dataset ecosystem treats it as a valid user name input. 

* As an analyst, you will have to work with the original dataset, BUT you cant change or modify the original dataset. This is called **"Data Integrity"**, and it has to be maintained always. <mark>All you can do is figure out trends, patterns, give insights and do analysis.</mark> 

* **DOING ANALYSIS DOES NOT MEAN MODIFYING ORIGINAL DATA, IT HAS TO BE KEPT AS IS ALWAYS.**

* Ultimately SQL works with a database in which our data is stored. Now when you have to do some analysis, you will take a "screenshot" of data, means you will fetch the data or in other words you will copy that version of database at that point in time and run a bunch of code on it and drive some insights from it. <mark>BUT WE WILL NEVER BE MODIFYING THE ORIGINAL DATABASE. **We can fetch the data in any optimised way possible but in that process we cant make changes to the original dataset.**</mark> 

* **A majority of pre-processing on data that you will do in python, will be for running your machine learning models.** To run machine learning models, you have to use python for data pre-processing because SQL or any other tool wont be good for this purpose especially when massive datasets are there.

* **Data pre-processing taught here is just an introduction, a lot of it will be covered later.** 